File: src/lib/server/crypto/envelope.ts
/**
 * Penombre's sealed file format, v1. Mirrors `internal/envelope` in Go, and
 * `tests/fixtures/envelope-v1.json` holds vectors both must reproduce.
 *
 * 76-byte header (magic + version, key id, wrap nonce, file key wrapped by the
 * instance key with bytes 0..15 as AAD), then AES-256-GCM chunks of 64 KiB
 * whose nonce is the chunk index plus a last-chunk flag against truncation.
 */

import {
	createCipheriv,
	createDecipheriv,
	createHmac,
	randomBytes,
	timingSafeEqual,
} from "node:crypto";

export const HEADER_SIZE = 76;
export const CHUNK_SIZE = 64 * 1024;
export const TAG_SIZE = 16;
export const KEY_SIZE = 32;
const SEALED_CHUNK = CHUNK_SIZE + TAG_SIZE;
const MAGIC = Buffer.from("PNMBENC\x01", "latin1");

export interface Keyring {
	/** Seals new bytes. Absent: nothing is sealed, sealed files still open. */
	current?: Buffer;
	/** Retired keys, read-only. */
	previous: Buffer[];
}

export class CorruptEnvelopeError extends Error {
	constructor() {
		super("Sealed file is corrupt or truncated");
	}
}

export class KeyMissingError extends Error {
	constructor(readonly keyId: string) {
		super(`Sealed with key ${keyId}, not loaded (set ENCRYPTION_KEY)`);
	}
}

export function keyId(kek: Buffer): Buffer {
	return createHmac("sha256", kek)
		.update("penombre/key-id/v1")
		.digest()
		.subarray(0, 8);
}

/** The key matching `id`, and whether it is the current one. */
function findKey(
	keyring: Keyring,
	id: Uint8Array,
): { kek: Buffer; current: boolean } | undefined {
	const matches = (kek: Buffer) => timingSafeEqual(keyId(kek), id);
	if (keyring.current && matches(keyring.current)) {
		return { kek: keyring.current, current: true };
	}
	const old = keyring.previous.find(matches);
	return old ? { kek: old, current: false } : undefined;
}

export function isSealed(prefix: Uint8Array): boolean {
	return (
		prefix.byteLength >= MAGIC.length &&
		MAGIC.equals(prefix.subarray(0, MAGIC.length))
	);
}

export function sealedSize(plain: number): number {
	return (
		HEADER_SIZE + plain + TAG_SIZE * Math.max(1, Math.ceil(plain / CHUNK_SIZE))
	);
}

export function plainSize(sealed: number): number {
	const n = sealed - HEADER_SIZE;
	const chunks = Math.ceil(n / SEALED_CHUNK);
	if (n < TAG_SIZE || n - (chunks - 1) * SEALED_CHUNK < TAG_SIZE) {
		throw new CorruptEnvelopeError();
	}
	return n - TAG_SIZE * chunks;
}

export function sealHeader(
	kek: Buffer,
	dek: Buffer,
	nonce: Buffer = randomBytes(12),
): Buffer {
	const aad = Buffer.concat([MAGIC, keyId(kek)]);
	const cipher = createCipheriv("aes-256-gcm", kek, nonce);
	cipher.setAAD(aad);
	const wrapped = Buffer.concat([cipher.update(dek), cipher.final()]);
	return Buffer.concat([aad, nonce, wrapped, cipher.getAuthTag()]);
}

/** The file key inside a header, and whether the current key wrapped it. */
export function openHeader(
	keyring: Keyring,
	header: Uint8Array,
): { dek: Buffer; current: boolean } {
	const h = Buffer.from(header.buffer, header.byteOffset, header.byteLength);
	if (h.length < HEADER_SIZE || !isSealed(h)) {
		throw new CorruptEnvelopeError();
	}
	const found = findKey(keyring, h.subarray(8, 16));
	if (!found) {
		throw new KeyMissingError(h.subarray(8, 16).toString("hex"));
	}
	try {
		const decipher = createDecipheriv(
			"aes-256-gcm",
			found.kek,
			h.subarray(16, 28),
		);
		decipher.setAAD(h.subarray(0, 16));
		decipher.setAuthTag(h.subarray(60, HEADER_SIZE));
		const dek = Buffer.concat([
			decipher.update(h.subarray(28, 60)),
			decipher.final(),
		]);
		return { dek, current: found.current };
	} catch {
		throw new CorruptEnvelopeError();
	}
}

function chunkNonce(index: number, last: boolean): Buffer {
	const nonce = Buffer.alloc(12);
	nonce.writeBigUInt64BE(BigInt(index), 3);
	nonce[11] = last ? 1 : 0;
	return nonce;
}

function sealChunk(
	dek: Buffer,
	index: number,
	last: boolean,
	plain: Uint8Array,
): Buffer {
	const cipher = createCipheriv("aes-256-gcm", dek, chunkNonce(index, last));
	return Buffer.concat([
		cipher.update(plain),
		cipher.final(),
		cipher.getAuthTag(),
	]);
}

function openChunk(
	dek: Buffer,
	index: number,
	last: boolean,
	sealed: Uint8Array,
): Buffer {
	if (sealed.byteLength < TAG_SIZE) {
		throw new CorruptEnvelopeError();
	}
	const body = sealed.subarray(0, sealed.byteLength - TAG_SIZE);
	try {
		const decipher = createDecipheriv(
			"aes-256-gcm",
			dek,
			chunkNonce(index, last),
		);
		decipher.setAuthTag(sealed.subarray(sealed.byteLength - TAG_SIZE));
		return Buffer.concat([decipher.update(body), decipher.final()]);
	} catch {
		throw new CorruptEnvelopeError();
	}
}

/** Seal a whole buffer. `dek` and `nonce` are fixed only by test vectors. */
export function seal(
	kek: Buffer,
	plain: Uint8Array,
	dek: Buffer = randomBytes(KEY_SIZE),
	nonce?: Buffer,
): Buffer {
	const chunks = Math.max(1, Math.ceil(plain.byteLength / CHUNK_SIZE));
	const parts = [sealHeader(kek, dek, nonce)];
	for (let index = 0; index < chunks; index++) {
		parts.push(
			sealChunk(
				dek,
				index,
				index === chunks - 1,
				plain.subarray(index * CHUNK_SIZE, (index + 1) * CHUNK_SIZE),
			),
		);
	}
	return Buffer.concat(parts);
}

/**
 * Seals `source`, re-chunked to 64 KiB, one sealed chunk per pull: a slow
 * writer holds back the reads instead of queueing the whole file sealed.
 */
export function sealStream(
	kek: Buffer,
	source: ReadableStream<Uint8Array>,
): ReadableStream<Uint8Array> {
	const dek = randomBytes(KEY_SIZE);
	const reader = source.getReader();
	let buffered: Buffer = Buffer.alloc(0);
	let index = 0;
	let done = false;
	return new ReadableStream<Uint8Array>({
		start(controller) {
			controller.enqueue(sealHeader(kek, dek));
		},
		async pull(controller) {
			// A full chunk is only known not to be the last once a byte follows it.
			while (!done && buffered.length <= CHUNK_SIZE) {
				const next = await reader.read();
				if (next.done) {
					done = true;
				} else {
					const bytes = Buffer.from(
						next.value.buffer,
						next.value.byteOffset,
						next.value.byteLength,
					);
					buffered =
						buffered.length > 0 ? Buffer.concat([buffered, bytes]) : bytes;
				}
			}
			if (buffered.length > CHUNK_SIZE) {
				controller.enqueue(
					sealChunk(dek, index++, false, buffered.subarray(0, CHUNK_SIZE)),
				);
				buffered = buffered.subarray(CHUNK_SIZE);
				return;
			}
			controller.enqueue(sealChunk(dek, index, true, buffered));
			controller.close();
		},
		cancel(reason) {
			return reader.cancel(reason);
		},
	});
}

/** Where a sealed file's bytes come from, so a range read fetches only what it needs. */
export interface SealedSource {
	size: number;
	read: (offset: number, length: number) => Promise<Uint8Array>;
}

/** Validates the header and the last chunk, which is what catches truncation. */
async function openSource(keyring: Keyring, source: SealedSource) {
	const size = plainSize(source.size);
	const { dek } = openHeader(keyring, await source.read(0, HEADER_SIZE));
	const chunks = Math.max(1, Math.ceil(size / CHUNK_SIZE));
	const chunk = async (index: number) => {
		const last = index === chunks - 1;
		const length = last ? size - index * CHUNK_SIZE + TAG_SIZE : SEALED_CHUNK;
		return openChunk(
			dek,
			index,
			last,
			await source.read(HEADER_SIZE + index * SEALED_CHUNK, length),
		);
	};
	const tail = await chunk(chunks - 1);
	return { size, chunks, chunk, tail };
}

/**
 * Plaintext bytes `start..end` (inclusive) as a pull stream: one chunk is
 * read and opened per pull, so an open-ended tail is never decrypted ahead.
 */
export async function openRange(
	keyring: Keyring,
	source: SealedSource,
	{
		start = 0,
		end,
		onClose = () => undefined,
	}: { start?: number; end?: number; onClose?: () => void } = {},
): Promise<ReadableStream<Uint8Array>> {
	const opened = await openSource(keyring, source);
	const last = Math.min(end ?? opened.size - 1, opened.size - 1);
	let index = Math.floor(start / CHUNK_SIZE);
	const lastIndex = Math.floor(last / CHUNK_SIZE);
	let closed = false;
	const close = () => {
		if (!closed) {
			closed = true;
			onClose();
		}
	};
	return new ReadableStream<Uint8Array>({
		async pull(controller) {
			if (start > last || index > lastIndex) {
				controller.close();
				close();
				return;
			}
			try {
				const plain =
					index === opened.chunks - 1 ? opened.tail : await opened.chunk(index);
				const base = index * CHUNK_SIZE;
				controller.enqueue(
					plain.subarray(
						Math.max(0, start - base),
						Math.min(CHUNK_SIZE, last - base + 1),
					),
				);
				index++;
			} catch (error) {
				controller.error(error);
				close();
			}
		},
		cancel() {
			close();
		},
	});
}

export function openWhole(keyring: Keyring, sealed: Uint8Array): Buffer {
	const size = plainSize(sealed.byteLength);
	const { dek } = openHeader(keyring, sealed);
	const chunks = Math.max(1, Math.ceil(size / CHUNK_SIZE));
	// Unpooled, so `.buffer` is exactly the plaintext.
	const plain = Buffer.allocUnsafeSlow(size);
	for (let index = 0; index < chunks; index++) {
		const from = HEADER_SIZE + index * SEALED_CHUNK;
		plain.set(
			openChunk(
				dek,
				index,
				index === chunks - 1,
				sealed.subarray(from, Math.min(from + SEALED_CHUNK, sealed.byteLength)),
			),
			index * CHUNK_SIZE,
		);
	}
	return plain;
}

File: internal/envelope/envelope.go
// Package envelope reads and writes Penombre's sealed file format (v1): a
// 76-byte header carrying the file's own key wrapped by the instance key,
// then AES-256-GCM chunks of 64 KiB. Mirrors src/lib/server/crypto/envelope.ts.
package envelope

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"os"
	"sync"
)

const (
	HeaderSize  = 76
	ChunkSize   = 64 << 10
	TagSize     = 16
	KeySize     = 32
	sealedChunk = ChunkSize + TagSize
	magicSize   = 8
	idSize      = 8
	nonceSize   = 12
)

var magic = []byte("PNMBENC\x01")

var ErrCorrupt = errors.New("sealed file is corrupt or truncated")

// KeyMissingError names a key id no loaded key matches.
type KeyMissingError struct{ ID string }

func (e KeyMissingError) Error() string {
	return fmt.Sprintf("sealed with key %s, not loaded (set ENCRYPTION_KEY on the worker)", e.ID)
}

// Keyring is the instance key plus retired ones kept to read older files.
type Keyring struct {
	Current  []byte
	Previous [][]byte
}

// Enabled reports whether new bytes are sealed.
func (k Keyring) Enabled() bool { return len(k.Current) == KeySize }

func KeyID(kek []byte) []byte {
	mac := hmac.New(sha256.New, kek)
	mac.Write([]byte("penombre/key-id/v1"))
	return mac.Sum(nil)[:idSize]
}

// find returns the key matching id, and whether it is the current one.
func (k Keyring) find(id []byte) ([]byte, bool) {
	if k.Current != nil && subtle.ConstantTimeCompare(KeyID(k.Current), id) == 1 {
		return k.Current, true
	}
	for _, key := range k.Previous {
		if subtle.ConstantTimeCompare(KeyID(key), id) == 1 {
			return key, false
		}
	}
	return nil, false
}

// Has reports whether a key with this hex id is loaded.
func (k Keyring) Has(id string) bool {
	raw, err := hex.DecodeString(id)
	if err != nil {
		return false
	}
	key, _ := k.find(raw)
	return key != nil
}

var (
	defaultMu   sync.RWMutex
	defaultRing Keyring
)

// SetDefault installs the worker's keyring, once at boot.
func SetDefault(k Keyring) {
	defaultMu.Lock()
	defer defaultMu.Unlock()
	defaultRing = k
}

func Default() Keyring {
	defaultMu.RLock()
	defer defaultMu.RUnlock()
	return defaultRing
}

func IsSealed(prefix []byte) bool {
	return len(prefix) >= magicSize && bytes.Equal(prefix[:magicSize], magic)
}

// SniffFile reports whether the file at path starts with the envelope magic.
func SniffFile(path string) (bool, error) {
	f, err := os.Open(path)
	if err != nil {
		return false, err
	}
	defer f.Close()
	buf := make([]byte, magicSize)
	n, err := io.ReadFull(f, buf)
	if errors.Is(err, io.ErrUnexpectedEOF) || errors.Is(err, io.EOF) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return IsSealed(buf[:n]), nil
}

func SealedSize(plain int64) int64 {
	chunks := max(1, (plain+ChunkSize-1)/ChunkSize)
	return HeaderSize + plain + TagSize*chunks
}

func PlainSize(sealed int64) (int64, error) {
	n := sealed - HeaderSize
	if n < TagSize {
		return 0, ErrCorrupt
	}
	chunks := (n + sealedChunk - 1) / sealedChunk
	if n-(chunks-1)*sealedChunk < TagSize {
		return 0, ErrCorrupt
	}
	return n - TagSize*chunks, nil
}

func gcm(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

// SealHeader wraps dek under kek with the given 12-byte nonce.
func SealHeader(kek, dek, nonce []byte) ([]byte, error) {
	if len(kek) != KeySize || len(dek) != KeySize || len(nonce) != nonceSize {
		return nil, errors.New("envelope: bad key or nonce length")
	}
	aead, err := gcm(kek)
	if err != nil {
		return nil, err
	}
	h := make([]byte, 0, HeaderSize)
	h = append(h, magic...)
	h = append(h, KeyID(kek)...)
	h = append(h, nonce...)
	return aead.Seal(h, nonce, dek, h[:magicSize+idSize]), nil
}

// OpenHeader unwraps the file key; current is false for a retired key.
func (k Keyring) OpenHeader(h []byte) (dek []byte, current bool, err error) {
	if len(h) < HeaderSize || !IsSealed(h) {
		return nil, false, ErrCorrupt
	}
	id := h[magicSize : magicSize+idSize]
	kek, current := k.find(id)
	if kek == nil {
		return nil, false, KeyMissingError{ID: hex.EncodeToString(id)}
	}
	aead, err := gcm(kek)
	if err != nil {
		return nil, false, err
	}
	nonce := h[16:28]
	dek, err = aead.Open(nil, nonce, h[28:HeaderSize], h[:16])
	if err != nil {
		return nil, false, ErrCorrupt
	}
	return dek, current, nil
}

// Rewrap re-seals a header's file key under the current key, fresh nonce.
func (k Keyring) Rewrap(h []byte) ([]byte, error) {
	if !k.Enabled() {
		return nil, errors.New("envelope: no current key to rewrap with")
	}
	dek, _, err := k.OpenHeader(h)
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, nonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	return SealHeader(k.Current, dek, nonce)
}

func chunkNonce(i int64, last bool) []byte {
	n := make([]byte, nonceSize)
	binary.BigEndian.PutUint64(n[3:11], uint64(i))
	if last {
		n[11] = 1
	}
	return n
}

// Reader decrypts a sealed file with random access. Safe for concurrent use.
type Reader struct {
	src    io.ReaderAt
	aead   cipher.AEAD
	size   int64
	chunks int64

	mu     sync.Mutex
	cached int64
	plain  []byte
}

// Open validates the header of a sealed file of sealedSize bytes.
func (k Keyring) Open(src io.ReaderAt, sealedSize int64) (*Reader, error) {
	h := make([]byte, HeaderSize)
	if _, err := src.ReadAt(h, 0); err != nil {
		return nil, ErrCorrupt
	}
	dek, _, err := k.OpenHeader(h)
	if err != nil {
		return nil, err
	}
	size, err := PlainSize(sealedSize)
	if err != nil {
		return nil, err
	}
	aead, err := gcm(dek)
	if err != nil {
		return nil, err
	}
	r := &Reader{src: src, aead: aead, size: size, chunks: max(1, (size+ChunkSize-1)/ChunkSize), cached: -1}
	// The last chunk's flag is what catches truncation, so check it up front:
	// a range read that never reaches the end must still refuse a cut file.
	if err := r.chunk(r.chunks - 1); err != nil {
		return nil, err
	}
	return r, nil
}

func (r *Reader) Size() int64 { return r.size }

// chunk decrypts chunk i into the cache. Caller holds mu.
func (r *Reader) chunk(i int64) error {
	if r.cached == i {
		return nil
	}
	length := int64(sealedChunk)
	if i == r.chunks-1 {
		length = r.size - i*ChunkSize + TagSize
	}
	buf := make([]byte, length)
	if _, err := r.src.ReadAt(buf, HeaderSize+i*sealedChunk); err != nil && !errors.Is(err, io.EOF) {
		return err
	}
	plain, err := r.aead.Open(buf[:0], chunkNonce(i, i == r.chunks-1), buf, nil)
	if err != nil {
		r.cached = -1
		return ErrCorrupt
	}
	r.cached, r.plain = i, plain
	return nil
}

func (r *Reader) ReadAt(p []byte, off int64) (int, error) {
	if off >= r.size {
		return 0, io.EOF
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	n := 0
	for n < len(p) && off < r.size {
		i := off / ChunkSize
		if err := r.chunk(i); err != nil {
			return n, err
		}
		c := copy(p[n:], r.plain[off-i*ChunkSize:])
		n += c
		off += int64(c)
	}
	if n < len(p) {
		return n, io.EOF
	}
	return n, nil
}

// OpenFile opens path as plaintext: sealed files through k, others as is.
func (k Keyring) OpenFile(path string) (io.ReadSeekCloser, int64, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, 0, err
	}
	info, err := f.Stat()
	if err != nil {
		f.Close()
		return nil, 0, err
	}
	prefix := make([]byte, magicSize)
	if n, _ := f.ReadAt(prefix, 0); !IsSealed(prefix[:n]) {
		return f, info.Size(), nil
	}
	r, err := k.Open(f, info.Size())
	if err != nil {
		f.Close()
		return nil, 0, err
	}
	return sectionCloser{io.NewSectionReader(r, 0, r.Size()), f}, r.Size(), nil
}

type sectionCloser struct {
	*io.SectionReader
	io.Closer
}

// Writer seals everything written to it. Close writes the final chunk.
type Writer struct {
	dst  io.Writer
	aead cipher.AEAD
	buf  []byte
	i    int64
}

func NewWriter(dst io.Writer, kek []byte) (*Writer, error) {
	dek := make([]byte, KeySize)
	nonce := make([]byte, nonceSize)
	if _, err := rand.Read(dek); err != nil {
		return nil, err
	}
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	return NewWriterWith(dst, kek, dek, nonce)
}

// NewWriterWith takes the file key and wrap nonce; NewWriter picks them at
// random. Fixed values are for test vectors only.
func NewWriterWith(dst io.Writer, kek, dek, nonce []byte) (*Writer, error) {
	h, err := SealHeader(kek, dek, nonce)
	if err != nil {
		return nil, err
	}
	aead, err := gcm(dek)
	if err != nil {
		return nil, err
	}
	if _, err := dst.Write(h); err != nil {
		return nil, err
	}
	return &Writer{dst: dst, aead: aead, buf: make([]byte, 0, ChunkSize+1)}, nil
}

func (w *Writer) Write(p []byte) (int, error) {
	n := len(p)
	for len(p) > 0 {
		take := min(len(p), ChunkSize+1-len(w.buf))
		w.buf = append(w.buf, p[:take]...)
		p = p[take:]
		// A full chunk is only known not to be the last once a byte follows it.
		if len(w.buf) > ChunkSize {
			if err := w.flush(w.buf[:ChunkSize], false); err != nil {
				return 0, err
			}
			w.buf = append(w.buf[:0], w.buf[ChunkSize])
		}
	}
	return n, nil
}

func (w *Writer) flush(plain []byte, last bool) error {
	_, err := w.dst.Write(w.aead.Seal(nil, chunkNonce(w.i, last), plain, nil))
	w.i++
	return err
}

func (w *Writer) Close() error { return w.flush(w.buf, true) }

// ErrNoKey is sealing asked of a process that has no current key.
var ErrNoKey = errors.New("asked to seal, but no ENCRYPTION_KEY is loaded")

// SealTo seals what is written to the result into w when seal is set, else
// passes it through. Closing it ends the envelope, never w itself.
func (k Keyring) SealTo(w io.Writer, seal bool) (io.WriteCloser, error) {
	if !seal {
		return nopCloser{w}, nil
	}
	if !k.Enabled() {
		return nil, ErrNoKey
	}
	return NewWriter(w, k.Current)
}

type nopCloser struct{ io.Writer }

func (nopCloser) Close() error { return nil }
